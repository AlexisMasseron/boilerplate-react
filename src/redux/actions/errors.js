export const ADD_ERROR = "ADD_ERROR";
export const addError = err => ({
    type: ADD_ERROR,
    payload: err
});
