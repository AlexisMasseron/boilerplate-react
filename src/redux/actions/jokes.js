export const FETCH_JOKE = "FETCH_JOKE";
export const fetchJoke = () => ({
    type: FETCH_JOKE
});

export const SET_JOKE = "SET_JOKE";
export const setJoke = joke => ({
    type: SET_JOKE,
    payload: joke
});

export const START_FETCHING_JOKE = "START_FETCHING_JOKE";
export const startFetchingJoke = () => ({
    type: START_FETCHING_JOKE
});

export const END_FETCHING_JOKE = "END_FETCHING_JOKE";
export const endFetchingJoke = () => ({
    type: END_FETCHING_JOKE
});
