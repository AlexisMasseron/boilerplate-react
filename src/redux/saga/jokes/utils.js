import { call } from "redux-saga/effects";

function* fetchRandomJoke() {
    const { value } = yield call(
        get,
        "https://api.chucknorris.io/jokes/random"
    );
    return value;
}

const get = url => fetch(url).then(res => res.json());

export { fetchRandomJoke };
