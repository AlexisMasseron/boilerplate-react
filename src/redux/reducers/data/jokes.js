import { SET_JOKE } from "../../actions/jokes";

const initialState = {
    currentJoke: ""
};

export default (state = initialState, { type, payload }) => {
    switch (type) {
        case SET_JOKE:
            return {
                ...state,
                currentJoke: payload
            };
        default:
            return state;
    }
};
