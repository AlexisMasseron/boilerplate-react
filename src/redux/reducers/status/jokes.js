import { START_FETCHING_JOKE, END_FETCHING_JOKE } from "../../actions/jokes";

const initialState = {
    isLoading: false
};

export default (state = initialState, { type, payload }) => {
    switch (type) {
        case START_FETCHING_JOKE:
            return {
                ...state,
                isLoading: true
            };
        case END_FETCHING_JOKE:
            return {
                ...state,
                isLoading: false
            };
        default:
            return state;
    }
};
