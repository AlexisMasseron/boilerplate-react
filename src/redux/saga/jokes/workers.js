import { put, call } from "redux-saga/effects";

// Actions
import {
    startFetchingJoke,
    endFetchingJoke,
    setJoke
} from "../../actions/jokes";
import { addError } from "../../actions/errors";

// Utils
import { fetchRandomJoke } from "./utils";

function* fetchJokeSaga() {
    try {
        yield put(startFetchingJoke());
        const joke = yield call(fetchRandomJoke);
        yield put(setJoke(joke));
    } catch (err) {
        yield put(addError("Unable to fetch the joke."));
    } finally {
        yield put(endFetchingJoke());
    }
}

export { fetchJokeSaga };
