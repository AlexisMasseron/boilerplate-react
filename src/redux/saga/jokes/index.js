import { takeLatest, all } from "redux-saga/effects";

import { FETCH_JOKE } from "../../actions/jokes";

import { fetchJokeSaga } from "./workers";

function* watchFetchJoke() {
    yield takeLatest(FETCH_JOKE, fetchJokeSaga);
}

export default function*() {
    yield all([watchFetchJoke()]);
}
