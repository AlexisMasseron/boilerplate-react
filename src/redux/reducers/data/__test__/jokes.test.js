import { SET_JOKE } from "../../../actions/jokes";
import jokesReducer from "../jokes";

const testJoke = "TEST_SET_JOKE";

describe("jokesReducer", () => {
    let state = {};

    it("default state match", () => {
        expect(jokesReducer(undefined, {})).toMatchSnapshot();
    });

    describe(`On ${SET_JOKE}`, () => {
        beforeEach(() => {
            state = jokesReducer(
                { currentJoke: "" },
                { type: SET_JOKE, payload: testJoke }
            );
        });

        it("currentJoke=TEST_SET_JOKE", () => {
            expect(state.currentJoke).toEqual(testJoke);
        });

        it("state matches snapshot", () => {
            expect(state).toMatchSnapshot();
        });
    });

    describe("On default", () => {
        beforeEach(() => {
            state = jokesReducer(undefined, { type: undefined });
        });

        it("state matches snapshot", () => {
            expect(state).toMatchSnapshot();
        });
    });
});
