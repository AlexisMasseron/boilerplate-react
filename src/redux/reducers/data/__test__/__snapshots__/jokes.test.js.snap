// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jokesReducer On SET_JOKE state matches snapshot 1`] = `
Object {
  "currentJoke": "TEST_SET_JOKE",
}
`;

exports[`jokesReducer On default state matches snapshot 1`] = `
Object {
  "currentJoke": "",
}
`;

exports[`jokesReducer default state match 1`] = `
Object {
  "currentJoke": "",
}
`;
